src/jvm/main/org/jetbrains/kotlinx/lincheck/strategy/managed/ObjectTracker.kt

```diff
@@ -23,38 +50,448 @@ interface ObjectTracker {
+    ): ObjectEntry {
+        return ObjectEntry(objNumber, objHashCode, objDisplayNumber, objReference)
+    }
+
+    protected fun computeObjectDisplayNumber(obj: Any): Int {
+        // In the case of general-purpose model checking mode, the thread numeration starts from 0.
+        val offset = if (obj is Thread && executionMode == ExecutionMode.GENERAL_PURPOSE_MODEL_CHECKER) -1 else 0
+        val objClassKey = obj.getSpecialClassNameRepresentation()?.classKey ?: obj.javaClass
+        return perClassObjectNumeration.update(objClassKey, default = offset) { it + 1 }
+    }
+
+    override fun registerThread(threadId: Int, thread: Thread) {
+        registerObjectIfAbsent(thread)
+    }
+
+    override fun registerNewObject(obj: Any): ObjectEntry =
+        registerObject(ObjectKind.NEW, obj)
+
+    override fun registerExternalObject(obj: Any): ObjectEntry =
+        registerObject(ObjectKind.EXTERNAL, obj)
+
+    private fun registerObject(kind: ObjectKind, obj: Any): ObjectEntry {
+        check(!obj.isImmutable)
+        cleanup()
+        val entry = createObjectEntry(
+            objNumber = ++objectCounter,
+            objHashCode = System.identityHashCode(obj),
+            objDisplayNumber = computeObjectDisplayNumber(obj),
+            objReference = WeakIdentityReference(obj, referenceQueue),
+            kind = kind,
+        )
+        objectIndex.updateInplace(entry.objectHashCode, default = mutableListOf()) {
+            cleanup()
+            add(entry)
+        }
+        return entry
+    }
+
+    private fun getEntries(objHashCode: IdentityHashCode): List<ObjectEntry>? {
+        val entries = objectIndex[objHashCode] ?: return null
+        entries.cleanup()
+        if (entries.isEmpty()) {
+            objectIndex.remove(objHashCode)
+            return null
+        }
+        return entries
+    }
+
+    override operator fun get(id: ObjectID): ObjectEntry? {
+        val objNumber = id.getObjectNumber()
+        val objHashCode = id.getObjectHashCode()
+        val entries = getEntries(objHashCode) ?: return null
+        return entries.find { it.objectNumber == objNumber }
+    }
+
+    override operator fun get(obj: Any?): ObjectEntry? {
+        val objHashCode = System.identityHashCode(obj)
+        val entries = getEntries(objHashCode) ?: return null
+        return entries.find { it.objectReference.get() === obj }
+    }
+
+    override fun enumerateObjectEntries(): Sequence<ObjectEntry> =
+        objectIndex.values.asSequence().flatten()
+
+    override fun retain(predicate: (ObjectEntry) -> Boolean) {
+        objectIndex.values.retainAll { entries ->
+            entries.cleanup()
+            entries.retainAll(predicate)
+            entries.isNotEmpty()
+        }
+    }
+
+    override fun reset() {
+        objectCounter = 0
+        objectIndex.clear()
+        referenceQueue.clear()
+        perClassObjectNumeration.clear()
+    }
+
+    private fun cleanup() {
```

Comment:
What is the difference between reset, clear and clean?

Reply:
- `reset()` resets the state of the object tracker between invocations.

- `clear()` is standard method of maps that is called from `reset()` to clear all the internal data structures of the tracker. 

- `cleanup()` is a custom method that is used to remove object entries registered in the object tracker whose objects were garbage collected. 

I've added comments to the `cleanup` methods, hopefully they will help to understand the code better.